File: HyBidXSampleObjC/ConfigurationViewController.m
//
//  ConfigurationViewController.m
//  HyBidXSampleObjC
//
//  Created by Fares Ben Hamouda on 28.02.20.
//  Copyright © 2020 Fares Ben Hamouda. All rights reserved.
//

#import "ConfigurationViewController.h"
#import "Config.h"
#import <HyBidX/HyBidX.h>

@interface ConfigurationViewController ()

@property (weak, nonatomic) IBOutlet UITextField *textFieldSDKToken;
@property (weak, nonatomic) IBOutlet UITextField *textFieldPlacement;
@property (weak, nonatomic) IBOutlet UITextField *textFieldUserID;
@property (weak, nonatomic) IBOutlet UITextField *textFieldAge;
@property (weak, nonatomic) IBOutlet UIPickerView *pickerViewGender;

@end

@implementation ConfigurationViewController

NSArray* genders;

- (void)viewDidLoad {
    [super viewDidLoad];
    genders = [[NSArray alloc]initWithObjects:@"male", @"female", @"not set", nil];
    [self initialValues];
}

-(void) initialValues {
    Config* config = [Config sharedInstance];
    _textFieldSDKToken.text = config.sdkToken;
    _textFieldPlacement.text = config.placement;
    [_pickerViewGender selectRow:genders.count-1 inComponent:0 animated:false];
    
    NSDictionary* userData = [HyBidX userData];
    _textFieldUserID.text = userData[@"userID"];
    _textFieldAge.text = userData[@"age"];
    
    if (![userData[@"gender"] isEqual:@""]) {
        [_pickerViewGender selectRow:[genders indexOfObject:userData[@"gender"]] inComponent:0 animated:false];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    if (self.isMovingFromParentViewController) {
        
        Config* config = [Config sharedInstance];
        if (![_textFieldSDKToken.text isEqual:[Config sharedInstance].sdkToken]) {
            config.sdkToken = _textFieldSDKToken.text;
            [HyBidX configureWith:config.sdkToken completion:nil];
        }
        if (![_textFieldPlacement.text isEqual:[Config sharedInstance].placement]) {
            config.placement = _textFieldPlacement.text;
        }
        if (![_textFieldUserID.text isEqual:@""]) {
            [HyBidX identifyUserWith:_textFieldUserID.text];
        }
        if (![_textFieldAge.text isEqual:@""]) {
            [HyBidX setUserAgeWith: [_textFieldAge.text intValue]];
        }
        if ([_pickerViewGender selectedRowInComponent:0] != genders.count-1) {
            [HyBidX setUserGenderWith:genders[ [_pickerViewGender selectedRowInComponent:0]]];
        }
    }
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(nonnull UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return genders.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return genders[row];
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    [self.view endEditing:YES];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [self.view endEditing:YES];
    return YES;
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.view endEditing:YES];
}
@end

File: HyBidXSample/ConfigurationViewController.swift
//
//  ConfigurationViewController.swift
//  KwizzadRTASample
//
//  Created by Fares Ben Hamouda on 26.02.20.
//  Copyright © 2020 Fares Ben Hamouda. All rights reserved.
//

import UIKit
import KwizzadRTA

class ConfigurationViewController: UIViewController {

    @IBOutlet weak var textFieldSDKToken: UITextField!
    @IBOutlet weak var textFieldPlacement: UITextField!
    
    @IBOutlet weak var textFieldUserID: UITextField!
    @IBOutlet weak var textFieldAge: UITextField!
    @IBOutlet weak var pickerViewGender: UIPickerView!
    
    let genders = ["male", "female", "not set"]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        initialValues()
    }
    
    func initialValues() {
        textFieldSDKToken.text = sdkToken
        textFieldPlacement.text = placement
        pickerViewGender.selectRow(genders.count-1, inComponent: 0, animated: false)

        let userData: [String: Any] = KwizzadRTA.userData()
        
        if let userID = userData["userID"] as? String {
            textFieldUserID.text = userID
        }
        
        if let ageString = userData["age"] as? String {
            textFieldAge.text = ageString
        }
        
        if let gender = userData["gender"] as? String {
            pickerViewGender.selectRow(genders.firstIndex(of: gender) ?? genders.count-1, inComponent: 0, animated: false)
        }
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

        if self.isMovingFromParent {
            if let textFieldSDKToken = textFieldSDKToken.text, textFieldSDKToken != sdkToken {
                sdkToken = textFieldSDKToken
                KwizzadRTA.configure(with: sdkToken)
            }
            if let textFieldPlacement = textFieldPlacement.text, textFieldPlacement != "" {
                placement = textFieldPlacement
            }
            if let userID = textFieldUserID.text, userID != "" {
                KwizzadRTA.identifyUser(with: userID)
            }
            if let ageString = textFieldAge.text, let age = Int(ageString) {
                KwizzadRTA.setUserAge(with: age)
            }
            if pickerViewGender.selectedRow(inComponent: 0) != genders.count-1 {
                KwizzadRTA.setUserGender(with: genders[pickerViewGender.selectedRow(inComponent: 0)])
            }
        }
    }
}

extension ConfigurationViewController: UIPickerViewDataSource, UIPickerViewDelegate {
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return genders.count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return genders[row]
    }
    
}

extension ConfigurationViewController: UITextFieldDelegate {
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        self.view.endEditing(true)
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        self.view.endEditing(true)
        return true
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.view.endEditing(true)
    }
    
}
